tests/all/cli_tests/threads.wat

```diff
@@ -0,0 +1,55 @@
+(module
+  ;; As we have discussed, it makes sense to make the shared memory an import
+  ;; so that all
+  (import "" "memory" (memory $shmem 1 1 shared))
+  (import "wasi_snapshot_preview1" "fd_write"
+    (func $__wasi_fd_write (param i32 i32 i32 i32) (result i32)))
+  (import "wasi" "thread_spawn"
+    (func $__wasi_thread_spawn (param i32) (result i32)))
+
+  (func (export "_start")
+    (local $i i32)
+
+    ;; Print "Hello _start".
+    (call $print (i32.const 32) (i32.const 13))
+
+    ;; Print "Hello wasi_thread_start" in several threads.
+    (drop (call $__wasi_thread_spawn (i32.const 0)))
+    (drop (call $__wasi_thread_spawn (i32.const 0)))
+    (drop (call $__wasi_thread_spawn (i32.const 0)))
+
+    ;; Wasmtime has no `wait/notify` yet, so we just spin to allow the threads
+    ;; to do their work.
+    (local.set $i (i32.const 2000000))
+    (loop $again
+      (local.set $i (i32.sub (local.get $i) (i32.const 1)))
+      (br_if $again (i32.gt_s (local.get $i) (i32.const 0)))
+    )
```

Comment:
Technically this can be done today, but I think this might be best done as some sort of "join" API from the host which isn't currently specified for `wasi-threads`, so while wait/notify could be added here there's no way to actually wait until all the child threads have exited.

Reply:
Looking at this again, I think I can rewrite this with a "join" loop after the spawn calls: each thread will atomically increment a memory location and notify the main thread; the main thread waits in a loop and only exits once the read memory equals the number of threads we are waiting for. This fixup could come after this PR merges?

Reply:
Ended up doing this anyways in this PR to try to get CI to go green.